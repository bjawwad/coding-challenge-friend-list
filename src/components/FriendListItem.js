import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import styles from './FriendListItem.css';

class FriendListItem extends Component {

  constructor(props) {
    super(props);

    this.state = {
      gender: 'select'
    }
  }

  render() {
    const { gender } = this.props;
    return (
      <li className={styles.friendListItem}>
        <div className={styles.friendInfos}>
          <div><span>{this.props.name}</span></div>
          <div>
            <small>xx friends in common</small>
          </div>
        </div>

        <div className={styles.genderSelect}>
          <select
            value={gender}
            onChange={(e) => { this.onGenderChange(e) }}
          >
            <option key='select'>Select</option>
            <option key='Male'>Male</option>
            <option key='Female'>Female</option>
          </select>
        </div>

        <div className={styles.friendActions}>
          <button className={`starred btn btn-default ${styles.btnAction}`}
            onClick={() => this.props.starFriend(this.props.id)}>
            <i className={classnames('fa', {
              'fa-star': this.props.starred,
              'fa-star-o': !this.props.starred
            })} />
          </button>
          <button className={`btn btn-default ${styles.btnAction}`}
            onClick={() => this.props.deleteFriend(this.props.id)}>
            <i className="fa fa-trash" />
          </button>
        </div>
      </li>
    );
  }

  onGenderChange = (e) => {
    if(this.props && this.props.changeGender){
      this.props.changeGender(e.target.value, this.props.id);
    }
    
  }

}

FriendListItem.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  starred: PropTypes.bool,
  starFriend: PropTypes.func
};

export default FriendListItem
