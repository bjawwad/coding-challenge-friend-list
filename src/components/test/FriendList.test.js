import React from 'react';
import renderer from 'react-test-renderer';
import FriendList from '../FriendList';
import FriendListItem from '../FriendListItem';
import {mount} from 'enzyme';

function mockFriends() {
    return [
        {
            name: 'Theodore Roosevelt',
            starred: true,
            id: '10'
        },
        {
            name: 'Abraham Lincoln',
            starred: false,
            id: '11'
        },
        {
            name: 'George Washington',
            starred: false,
            id: '20'
        }
    ]
}

function mockActions () {
    return {
        addFriend: jest.fn(),
        deleteFriend: jest.fn(),
        starFriend: jest.fn(),
        changeGender: jest.fn()
    }
}

describe('<FriendList />', () => {
    it('should render without crashing', () => {
        const friends = mockFriends();
        const actions = mockActions();
        const wrapper = renderer.create(<FriendList friends={friends} actions={actions}/>).toJSON();
        expect(wrapper).toMatchSnapshot();
    })

    it('should render three friends', () => {
        const friends = mockFriends();
        const actions = mockActions();
        const wrapper = mount(<FriendList friends={friends} actions={actions}/>);
        const friendItems = wrapper.find('ul').children().map(child => child.childAt(0));
        expect(friendItems).toHaveLength(3);
    })
})