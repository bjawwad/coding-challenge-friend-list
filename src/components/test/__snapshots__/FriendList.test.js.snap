exports[`<FriendList /> should render without crashing 1`] = `
<ul
  className={undefined}>
  <li
    className={undefined}>
    <div
      className={undefined}>
      <div>
        <span>
          Theodore Roosevelt
        </span>
      </div>
      <div>
        <small>
          xx friends in common
        </small>
      </div>
    </div>
    <div
      className={undefined}>
      <select
        onChange={[Function]}
        value={undefined}>
        <option>
          Select
        </option>
        <option>
          Male
        </option>
        <option>
          Female
        </option>
      </select>
    </div>
    <div
      className={undefined}>
      <button
        className="starred btn btn-default undefined"
        onClick={[Function]}>
        <i
          className="fa fa-star" />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}>
        <i
          className="fa fa-trash" />
      </button>
    </div>
  </li>
  <li
    className={undefined}>
    <div
      className={undefined}>
      <div>
        <span>
          Abraham Lincoln
        </span>
      </div>
      <div>
        <small>
          xx friends in common
        </small>
      </div>
    </div>
    <div
      className={undefined}>
      <select
        onChange={[Function]}
        value={undefined}>
        <option>
          Select
        </option>
        <option>
          Male
        </option>
        <option>
          Female
        </option>
      </select>
    </div>
    <div
      className={undefined}>
      <button
        className="starred btn btn-default undefined"
        onClick={[Function]}>
        <i
          className="fa fa-star-o" />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}>
        <i
          className="fa fa-trash" />
      </button>
    </div>
  </li>
  <li
    className={undefined}>
    <div
      className={undefined}>
      <div>
        <span>
          George Washington
        </span>
      </div>
      <div>
        <small>
          xx friends in common
        </small>
      </div>
    </div>
    <div
      className={undefined}>
      <select
        onChange={[Function]}
        value={undefined}>
        <option>
          Select
        </option>
        <option>
          Male
        </option>
        <option>
          Female
        </option>
      </select>
    </div>
    <div
      className={undefined}>
      <button
        className="starred btn btn-default undefined"
        onClick={[Function]}>
        <i
          className="fa fa-star-o" />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}>
        <i
          className="fa fa-trash" />
      </button>
    </div>
  </li>
</ul>
`;
